xde/src/route.rs

```diff
@@ -0,0 +1,587 @@
+/// 2304 |@                                        14791
+/// 2560 |@@                                       32901
+/// 2816 |@                                        10730
+/// 3072 |                                         3459
+/// ```
+///
+/// Naturally, bringing this down to O(ns) is desirable. Usually, illumos
+/// holds `ire_t`s per `conn_t`, but we're aiming to be more fine-grained
+/// with DDM -- so we need a tradeoff between 'asking about the best route
+/// per-packet' and 'holding a route until it is expired'. We choose, for noe,
+/// to hold a route for 100ms.
+///
+/// Note, this uses a `BTreeMap`, but we would prefer the more consistent
+/// (faster) add/remove costs of a `HashMap`. As `BTreeMap` modification costs
+/// outpace the cost of `next_hop` between 256--512 entries, we currently set 512
+/// as a cap on cache size to prevent significant packet stalls. This may be tricky
+/// to tune.
+///
+/// (See: https://github.com/oxidecomputer/opte/pull/499#discussion_r1581164767
+/// for some performance numbers.)
+#[derive(Clone)]
+pub struct RouteCache(Arc<KRwLock<BTreeMap<RouteKey, CachedRoute>>>);
+
+impl Default for RouteCache {
+    fn default() -> Self {
+        let mut lock = KRwLock::new(BTreeMap::new());
+        lock.init(KRwLockType::Driver);
+        Self(lock.into())
+    }
+}
+
+impl RouteCache {
+    /// Retrieve a [`Route`] (device and L2 information) for a given `key`.
+    ///
+    /// This will retrieve an existing entry, if one exists from a recent
+    /// query, or computes the current route using `next_hop` on miss or
+    /// discovery of a stale entry.
+    pub fn next_hop<'b>(&self, key: RouteKey, xde: &'b XdeDev) -> Route<'b> {
+        let t = Moment::now();
+
+        let maybe_route = {
+            let route_cache = self.0.read();
+            route_cache.get(&key).copied()
+        };
+
+        match maybe_route {
+            Some(route) if route.is_valid(t) => return route.into_route(xde),
+            _ => {}
+        }
+
+        // Cache miss: intent is to now ask illumos, then insert.
+        let mut route_cache = self.0.write();
+        let space_remaining = route_cache.len() < MAX_CACHE_ENTRIES;
+
+        // Someone else may have written while we were taking the lock.
+        // DO NOT waste time if there's a good route.
+        let maybe_route = route_cache.entry(key);
+        let entry_exists = match &maybe_route {
+            Entry::Occupied(e) if e.get().is_valid(t) => {
+                return e.get().into_route(xde);
+            }
+            Entry::Occupied(_) => true,
+            _ => false,
+        };
+
+        // We've had a definitive flow miss, but we need to cap the cache
+        // size to prevent excessive modification latencies at high flow
+        // counts.
+        // If full and we have no old entry to update, drop the lock and do
+        // not insert.
+        // XXX: Want to profile in future to see if LRU expiry is
```

Comment:
I think this is worth an issue and measuring cache performance there is 1) a pool of destination route keys larger than the cap that is b) churning within the expiry time. One could easily imagine this scenarios arising for a busy web server with lots of clients. I think the reason we have the L4 hash in the key is to pin flows to paths, so expressing this as an LPM to reduce table size would probably be tricky?

Reply:
I'll add some extra probes around cache events so we can observe cases like this.

The L4 Hash in the key captures behaviour we have today, which is that OPTE uses multiple sled-to-sled paths depending on flow hash. We could recast that behaviour into an LPM of O(sleds) entries: if we can query and cache *all* reachability and preference information for each underlay destination (e.g., over both NIC links), then we can handle distribution over paths ourselves. We could possibly walk `ire->ire_bucket` in the same vein as `ire_round_robin` to fill that out?

I think that would be the right call; it'd be followup work that would likely replace large parts of this, so it's up to you on whether we use what we have for now and keep an LPM in mind as the next step (until we can offload, that is 😄).

Reply:
Let's take the win in front of us and continue to consider recasting things to fit in an LPM lookup later on.

Reply:
Thanks again, #539 opened.